apps/wallet/src/background/accounts/zklogin/ZkLoginAccount.ts

```diff
@@ -282,12 +342,63 @@ export class ZkLoginAccount
 			randomness: randomness.toString(),
 			jwt,
 		};
-		ephemeralValue[serializeNetwork(activeNetwork)] = credentialsData;
-		await this.setEphemeralValue(ephemeralValue);
+		await this.saveCredentialsData(credentialsData);
 		await this.onUnlocked();
+
+		// On re-auth, we check if the account for the complementary
+		// legacy/non-legacy address needs to be imported. Additionally, if the
+		// complementary account has been imported before and is unlocked, we
+		// update its credentials. This sync does not need to block the login
+		// process.
+		this.#syncAlternateAccount(jwt, credentialsData);
+
 		return credentialsData;
 	}
 
+	async #syncAlternateAccount(jwt: string, credentialsData: CredentialData) {
+		const currentAddress = await this.address;
+		const salt = await fetchSalt(jwt);
+		const legacyAddress = jwtToAddress(jwt, salt, true);
+		const nonLegacyAddress = jwtToAddress(jwt, salt, false);
+
+		if (normalizeSuiAddress(legacyAddress) !== normalizeSuiAddress(currentAddress)) {
+			await this.#importOrUpdateAlternateAccount(legacyAddress, credentialsData);
+		}
+		if (normalizeSuiAddress(nonLegacyAddress) !== normalizeSuiAddress(currentAddress)) {
+			await this.#importOrUpdateAlternateAccount(nonLegacyAddress, credentialsData);
+		}
+	}
+
+	// TODO XXX FIXME: update alternate account credentials
+	async #importOrUpdateAlternateAccount(address: string, _credentialsData: CredentialData) {
+		const imported = await getAccountsByAddress(address);
+		if (imported.length) {
+			const [existing] = imported;
+			if (existing instanceof ZkLoginAccount && !(await existing.isLocked())) {
+				existing.saveCredentialsData(_credentialsData);
+			}
+		} else {
+			const cached = await this.getCachedData();
+			await addNewAccounts([
+				{
+					address,
+					createdAt: Date.now(),
+					publicKey: cached.publicKey,
+					selected: false,
+					type: 'zkLogin',
+					nickname: cached.nickname ? `${cached.nickname} (address 2)` : null,
+					lastUnlockedOn: null,
+				},
+			]);
```

Comment:
Do we need something like this here too?

```suggestion
			const [newAccount] = await addNewAccounts([
				{
					address,
					createdAt: Date.now(),
					publicKey: cached.publicKey,
					selected: false,
					type: 'zkLogin',
					nickname: cached.nickname ? `${cached.nickname} (address 2)` : null,
					lastUnlockedOn: null,
				},
			]);
			
			if (existing instanceof ZkLoginAccount && !(await existing.isLocked())) {
				existing.saveCredentialsData(_credentialsData);
			}
```